clippy_lints/src/utils/mod.rs

```diff
@@ -90,7 +91,15 @@ pub fn in_constant(cx: &LateContext<'_, '_>, id: HirId) -> bool {
 
 /// Returns `true` if this `expn_info` was expanded by any macro.
 pub fn in_macro(span: Span) -> bool {
-    span.ctxt().outer().expn_info().is_some()
+    if let Some(info) = span.ctxt().outer().expn_info() {
+        if let ExpnFormat::CompilerDesugaring(..) = info.format {
+            false
```

Comment:
Do we actually want this in the macro check? As you can see it inadvertently fixed a bug in `into_iter_on_ref.rs`, but there may be cases where we don't care about desugarings. Perhaps we should just have both kinds of checks.

The nonminimal bool test fails without this because the visitor bails when it sees the if desugaring.

Reply:
we may be able to do this more fine-grained in the future by passing in more than a span (e.g. a `CompilerDesugaring` variant that we allow). For now it seems ok, since this can only happen in HIR anyway, and our HIR lints already know about the compiler expansions they care about